### PR TITLE
Track PRIMARY KEY column ordering with a sql.PrimaryKeySchema struct.

### PR DESCRIPTION
**Pull request by max-hoffman**
The parser exposes PRIMARY KEYs in the TableSpec for `CREATE TABLE`,
which we pluck before planning. Integrators expected to use the new
`sql.PrimaryKeySchema` even for keyless tables.

`CREATE TABLE LIKE` and `CREATE TABLE SELECT` are handled separately
and need more work to support the same order tracking.

`sql.PrimaryKeyAlterableTable` now includes `PrimaryKeySchema()` and `Pks()`
methods to access the new schema info.

### CODE REVIEW

File: sql/plan/update_source.go [hunk 1]
```diff
@@ -20,7 +20,7 @@ import (
 	"github.com/dolthub/go-mysql-server/sql"
 )
 
-// UpdateSource is the source of updates for an Update node. Its schema is the concatenation of the old and new rows,
+// UpdateSource is the source of updates for an Update node. Its pkSchema is the concatenation of the old and new rows,
```
Comment (zachmu) [inline sql/plan/update_source.go:23]: Rename operation got too aggressive, would revert and try again

File: sql/core.go [hunk 2]
```diff
@@ -358,6 +358,9 @@ type PrimaryKeyAlterableTable interface {
 	CreatePrimaryKey(ctx *Context, columns []IndexColumn) error
 	// DropPrimaryKey drops a primary key on a table. Returns an error if that table does not have a key.
 	DropPrimaryKey(ctx *Context) error
+	// PkOrdinals returns a list of the primary key column orderings
+	// ex: `CREATE TABLE a (x int, y int PRIMARY KEY (y,x)` has columns `[x,y]`,  primary key `[y,x]`, and pk ordinals `[1,0]`
+	PkOrdinals() []int
```
Comment (zachmu) [inline sql/core.go:363]: This should be symmetrical with Create, return []IndexColumn

File: sql/schema.go [hunk 3]
```diff
@@ -103,3 +103,8 @@ func IsKeyless(s Schema) bool {
 
 	return true
 }
+
+type PkSchema struct {
+	Schema
+	PkOrdinals []int
```
Comment (zachmu) [inline sql/schema.go:109]: Not suitable for a public interface. Just use a slice of column names.

File: sql/schema.go [hunk 4]
```diff
@@ -103,3 +103,8 @@ func IsKeyless(s Schema) bool {
 
 	return true
 }
+
+type PkSchema struct {
```
Comment (zachmu) [inline sql/schema.go:107]: Use a full name here, `PrimaryKeySchema`

File: sql/plan/ddl.go [hunk 5]
```diff
@@ -90,15 +90,15 @@ func (i *IndexDefinition) String() string {
 	return i.IndexName
 }
 
-// TableSpec is a node describing the schema of a table.
+// TableSpec is a node describing the pkSchema of a table.
 type TableSpec struct {
-	Schema  sql.Schema
+	Schema  *sql.PkSchema
```
Comment (zachmu) [inline sql/plan/ddl.go:95]: No real reason to use a pointer type here or elsewhere

**Review by zachmu — COMMENTED** [review]
Initial comments on names and interfaces

File: sql/plan/show_create_table_test.go [hunk 6]
```diff
@@ -33,13 +33,13 @@ func TestShowCreateTable(t *testing.T) {
...
 		"test-table",
-		sql.Schema{
+		sql.NewPrimaryKeySchema(sql.Schema{
 			&sql.Column{Name: "baz", Type: sql.Text, Default: nil, Nullable: false, PrimaryKey: true},
 			&sql.Column{Name: "zab", Type: sql.Int32, Default: parse.MustStringToColumnDefaultValue(ctx, "0", sql.Int32, true), Nullable: true, PrimaryKey: true},
 			&sql.Column{Name: "bza", Type: sql.Uint64, Default: parse.MustStringToColumnDefaultValue(ctx, "0", sql.Uint64, true), Nullable: true, Comment: "hello"},
 			&sql.Column{Name: "foo", Type: sql.MustCreateStringWithDefaults(sqltypes.VarChar, 123), Default: nil, Nullable: true},
 			&sql.Column{Name: "pok", Type: sql.MustCreateStringWithDefaults(sqltypes.Char, 123), Default: nil, Nullable: true},
-		})
+		}, []int{1}))
```
Comment (zachmu) [inline sql/plan/show_create_table_test.go:42]: This doesn't match the above

Comment (zachmu) [reply sql/plan/show_create_table_test.go:42]: Probably similar errors in other test files, I have not checked exhaustively

File: enginetest/enginetests.go [hunk 7]
```diff
@@ -1786,6 +1786,170 @@ func TestCreateDatabase(t *testing.T, harness Harness) {
 	})
 }
 
+func TestPkOrdinals(t *testing.T, harness Harness) {
+	e := NewEngine(t, harness)
+	ctx := NewContext(harness)
+
+	TestQuery(t, harness, e, "CREATE DATABASE testdb", []sql.Row{{sql.OkResult{RowsAffected: 1}}}, nil, nil)
```
Comment (zachmu) [inline enginetest/enginetests.go:1793]: Probably don't want to create DBs like this if you can avoid it, just use myDb like other tests

File: enginetest/enginetests.go [hunk 8]
```diff
@@ -1786,6 +1786,170 @@ func TestCreateDatabase(t *testing.T, harness Harness) {
...
+
+	db, err := e.Analyzer.Catalog.Database("testdb")
+	require.NoError(t, err)
+
+	TestQuery(t, harness, e, "USE testdb", []sql.Row(nil), nil, nil)
+
+	require.Equal(t, ctx.GetCurrentDatabase(), "testdb")
+
+	t.Run("CREATE table out of order PKs", func(t *testing.T) {
+		TestQuery(t, harness, e, "CREATE TABLE a (x int, y int, primary key (y,x))", []sql.Row(nil), nil, nil)
```
Comment (zachmu) [inline enginetest/enginetests.go:1803]: For queries like this where you aren't interested in asserting the result rows, use RunQuery instead

File: sql/schema.go [hunk 9]
```diff
@@ -103,3 +103,21 @@ func IsKeyless(s Schema) bool {
...
+type PrimaryKeySchema struct {
+	Schema
+	pkOrdinals []int
+}
+
+func NewPrimaryKeySchema(s Schema, pkOrds []int) PrimaryKeySchema {
+	return PrimaryKeySchema{Schema: s, pkOrdinals: pkOrds}
+}
+
+func (pks PrimaryKeySchema) WithOrdinals(o []int) PrimaryKeySchema {
```
Comment (zachmu) [inline sql/schema.go:116]: No real reason for this, you can just export pkOrdinals

File: sql/schema.go [hunk 10]
```diff
@@ -103,3 +103,21 @@ func IsKeyless(s Schema) bool {
...
+	pks.pkOrdinals = o
+	return pks
+}
+
+func (pks PrimaryKeySchema) PkOrdinals() []int {
```
Comment (zachmu) [inline sql/schema.go:121]: Same as above

File: sql/schema.go [hunk 11]
```diff
@@ -103,3 +103,21 @@ func IsKeyless(s Schema) bool {
 
 	return true
 }
+
+type PrimaryKeySchema struct {
```
Comment (zachmu) [inline sql/schema.go:107]: Need a type comment about where this is required / how it's used

File: sql/schema.go [hunk 12]
```diff
@@ -103,3 +103,21 @@ func IsKeyless(s Schema) bool {
...
+	Schema
+	pkOrdinals []int
+}
+
+func NewPrimaryKeySchema(s Schema, pkOrds []int) PrimaryKeySchema {
```
Comment (zachmu) [inline sql/schema.go:112]: Suggest making pkOrds a `...` varargs param, and change the behavior to automatically look for primary keys in the schema given if no ords are provided. Then you can avoid passing []int{} in all the test files.

Alternately, create a separate method, such as `PKSchemaFromSchema(schema)` that behaves the way I'm describing. And document how the method behaves in a comment.

File: sql/core.go [hunk 13]
```diff
@@ -358,6 +358,10 @@ type PrimaryKeyAlterableTable interface {
 	CreatePrimaryKey(ctx *Context, columns []IndexColumn) error
 	// DropPrimaryKey drops a primary key on a table. Returns an error if that table does not have a key.
 	DropPrimaryKey(ctx *Context) error
+	// Pks returns a list of ordered primary key columns
+	Pks() []IndexColumn
+	// PrimaryKeySchema returns this table's PrimaryKeySchema
+	PrimaryKeySchema() PrimaryKeySchema
```
Comment (zachmu) [inline sql/core.go:364]: As mentioned elsewhere, this is defined on the wrong interface. Make a new one, `PrimaryKeyTable`

**Review by zachmu — COMMENTED** [review]
Implementation looks pretty good, but I have some comments on the interface

File: sql/core.go [hunk 14]
```diff
@@ -358,6 +358,10 @@ type PrimaryKeyAlterableTable interface {
 	CreatePrimaryKey(ctx *Context, columns []IndexColumn) error
 	// DropPrimaryKey drops a primary key on a table. Returns an error if that table does not have a key.
 	DropPrimaryKey(ctx *Context) error
+	// Pks returns a list of ordered primary key columns
+	Pks() []IndexColumn
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply sql/core.go:362]: Convenience method is handy, avoid running a for loop to pluck Pks every time we want them

File: sql/plan/show_create_table_test.go (hunk 6 above)
Comment (max-hoffman) [reply sql/plan/show_create_table_test.go:42]: i'll do another lookthrough to try and catch stragglers

File: enginetest/enginetests.go (hunk 8 above)
Comment (max-hoffman) [reply enginetest/enginetests.go:1803]: 👍

File: enginetest/enginetests.go (hunk 7 above)
Comment (max-hoffman) [reply enginetest/enginetests.go:1793]: 👍

File: sql/schema.go (hunk 12 above)
Comment (max-hoffman) [reply sql/schema.go:112]: f

Comment (max-hoffman) [reply sql/schema.go:112]: not making that mistake again

File: sql/schema.go [hunk 15]
```diff
@@ -103,3 +103,21 @@ func IsKeyless(s Schema) bool {
...
 	return true
 }
+
+// PrimaryKeySchema defines table metadata for columns and primary key ordering
+type PrimaryKeySchema struct {
+	Schema
+	PkOrdinals []int
+}
+
+func NewPrimaryKeySchema(s Schema, pkOrds ...int) PrimaryKeySchema {
```
Comment (zachmu) [inline sql/schema.go:113]: Document this behavior

**Review by zachmu — APPROVED** [review]

**Pull request merged**